src/extensionProvider.ts

```diff
@@ -93,65 +93,39 @@ export class ExtensionProvider implements IDappProvider {
+  async signTransaction<T extends ITransaction>(transaction: T): Promise<T> {
+    let response = await this.signTransactions([transaction]);
+    return response[0];
   }
 
-  async signTransaction(transaction: ITransaction): Promise<ISignedTransaction> {
-    const txResponse = await this.startBgrMsgChannel("signTransactions", {
+  async signTransactions<T extends ITransaction>(transactions: Array<T>): Promise<Array<T>> {
+    let extensionResponse = await this.startBgrMsgChannel("signTransactions", {
       from: this.account.address,
-      transactions: [transaction.toPlainObject()],
+      transactions: transactions.map(transaction => transaction.toPlainObject()),
     });
-    return this.getTransactionFactory().fromPlainObject(txResponse[0]);
-  }
 
-  async signTransactions(
-    transactions: Array<ITransaction>
-  ): Promise<Array<ITransaction>> {
-    transactions = transactions.map((transaction) =>
-      transaction.toPlainObject()
-    );
-    let txResponse = await this.startBgrMsgChannel("signTransactions", {
-      from: this.account.address,
-      transactions: transactions,
-    });
     try {
-      // TODO: Find out whether transaction.applySignature() is better suited here,
-      // It seems that the extension only sets "sender" and "signature" 
-      // (version, nonce, chainID, gasLimit etc. do not seem to ever be overridden).
-      // If so, we don't need factories here.
-      txResponse = txResponse.map((transaction: any) =>
-        this.getTransactionFactory().fromPlainObject(transaction)
-      );
-    } catch (error) {
-      throw new Error("Transaction canceled.");
+      for (let i = 0; i < transactions.length; i++) {
+        let transaction = transactions[i];
+        let plainSignedTransaction = extensionResponse[i];
+
+        transaction.applySignature(new Signature(plainSignedTransaction.signature), new Address(this.account.address));
+      }
+    } catch (error: any) {
+      throw new Error(`Transaction canceled: ${error.message}.`);
     }
 
-    return txResponse;
+    return transactions;
   }
 
-  async signMessage(message: ISignableMessage): Promise<ISignedMessage> {
+  async signMessage<T extends ISignableMessage>(message: T): Promise<T> {
     const data = {
       account: this.account.address,
-      // TODO: Why not message.serializeForSigningRaw()?
-      message: message.message.toString(),
+      message: message.message.toString()
     };
-    const signResponse = await this.startBgrMsgChannel("signMessage", data);
-    // TODO: Find out whether message.applySignature() is better suited here.
-    // If so, we don't need factories here.
-    const signedMsg = this.getMessageFactory().fromPlainObject({
-      // TODO: Why not signResponse.address?
-      address: data.account,
-      message: Buffer.from(signResponse.message),
-      signature: signResponse.signature
-    });
-
-    return signedMsg;
+    const extensionResponse = await this.startBgrMsgChannel("signMessage", data);
```

Comment:
Should we wrap this in a try catch, to avoid throwing back at the user any unhandled exceptions?

Reply:
Indeed, we should. Is is OK to postpone the improved error handling mechanism to a separate PR?